homeassistant/components/kaleidescape/media_player.py

```diff
@@ -0,0 +1,217 @@
+"""Kaleidescape Media Player."""
+
+from __future__ import annotations
+
+import logging
+from typing import TYPE_CHECKING
+
+from kaleidescape import const as kaleidescape_const
+
+from homeassistant.components.media_player import MediaPlayerEntity
+from homeassistant.components.media_player.const import (
+    SUPPORT_NEXT_TRACK,
+    SUPPORT_PAUSE,
+    SUPPORT_PLAY,
+    SUPPORT_PREVIOUS_TRACK,
+    SUPPORT_STOP,
+    SUPPORT_TURN_OFF,
+    SUPPORT_TURN_ON,
+)
+from homeassistant.const import STATE_IDLE, STATE_OFF, STATE_PAUSED, STATE_PLAYING
+from homeassistant.core import callback
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.util.dt import utcnow
+
+from .const import DOMAIN as KALEIDESCAPE_DOMAIN, NAME as KALEIDESCAPE_NAME
+
+if TYPE_CHECKING:
+    from datetime import datetime
+
+    from kaleidescape import Device as KaleidescapeDevice
+
+    from homeassistant.config_entries import ConfigEntry
+    from homeassistant.core import HomeAssistant
+    from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+
+SUPPORTED_FEATURES = (
+    SUPPORT_TURN_ON
+    | SUPPORT_TURN_OFF
+    | SUPPORT_PLAY
+    | SUPPORT_PAUSE
+    | SUPPORT_STOP
+    | SUPPORT_NEXT_TRACK
+    | SUPPORT_PREVIOUS_TRACK
+)
+
+KALEIDESCAPE_DEVICE_EVENTS = [
+    kaleidescape_const.DEVICE_POWER_STATE,
+    kaleidescape_const.FRIENDLY_NAME,
+    kaleidescape_const.PLAY_STATUS,
+    kaleidescape_const.MOVIE_LOCATION,
+    kaleidescape_const.SCREEN_MASK,
+    kaleidescape_const.VIDEO_COLOR,
+]
+
+KALEIDESCAPE_PLAYING_STATES = [
+    kaleidescape_const.PLAY_STATUS_PLAYING,
+    kaleidescape_const.PLAY_STATUS_FORWARD,
+    kaleidescape_const.PLAY_STATUS_REVERSE,
+]
+
+KALEIDESCAPE_PAUSED_STATES = [kaleidescape_const.PLAY_STATUS_PAUSED]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the platform from a config entry."""
+    entities = [KaleidescapeMediaPlayer(hass.data[KALEIDESCAPE_DOMAIN][entry.entry_id])]
+    async_add_entities(entities, True)
+
+
+class KaleidescapeMediaPlayer(MediaPlayerEntity):
+    """Representation of a Kaleidescape device."""
+
+    def __init__(self, device: KaleidescapeDevice) -> None:
+        """Initialize media player."""
+        self._device = device
+
+        self._attr_should_poll = False
+        self._attr_unique_id = device.serial_number
+        self._attr_name = f"{KALEIDESCAPE_NAME} {device.system.friendly_name}"
+        self._attr_supported_features = SUPPORTED_FEATURES
+
+    async def async_added_to_hass(self) -> None:
+        """Register update listener."""
+
+        @callback
+        def _update(event: str) -> None:
+            """Handle device state changes."""
+            self.async_write_ha_state()
+
+        self.async_on_remove(self._device.dispatcher.connect(_update).disconnect)
```

Comment:
It looks like this should be moved into a base class for the entity so it can be shared between platforms later on

Reply:
Great tip. Created KaleidescapeEntity file and class.